pkg/runner/healthcheck_common.go

```diff
@@ -0,0 +1,283 @@
+package runner
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/docker/go-units"
+	"github.com/ipfs/testground/pkg/api"
+	"github.com/ipfs/testground/pkg/docker"
+	"github.com/ipfs/testground/pkg/rpc"
+)
+
+type Checker func() (bool, error)
+type Fixer func() error
+
+// HealthcheckHelper is a strategy interface for runners.
+// Each runner may have required elements -- infrastructure, etc. which should be checked prior to
+// running plans. Individual checks are registered to the HealthcheckHelper using the Enlist()
+// method. With all of the checks enlisted, execute the checks, and optionally fixes, using the
+// RunChecks() method. The details of how the checks are performed is implementation specific.
+// Typically, the checker and fixer passed to the enlist method will be closures. These methods will
+// be called when RunChecks is executed.
+type HealthcheckHelper interface {
+	Enlist(name string, c Checker, f Fixer)
+	RunChecks(ctx context.Context, fix bool) error
+}
+
+type toDoElement struct {
+	Name    string
+	Checker Checker
+	Fixer   Fixer
+}
+
+// SequentialHealthcheckHelper implements HealthcheckHelper. Runchecks runs each check and fix
+// sequentially, in the order they are Enlist()'ed.
+type SequentialHealthcheckHelper struct {
+	toDo   []*toDoElement
+	report *api.HealthcheckReport
+}
+
+func (hh *SequentialHealthcheckHelper) Enlist(name string, c Checker, f Fixer) {
+	hh.toDo = append(hh.toDo, &toDoElement{name, c, f})
+}
+
+func (hh *SequentialHealthcheckHelper) RunChecks(ctx context.Context, fix bool) error {
+	for _, li := range hh.toDo {
+		// Check succeeds.
+		succeed, err := li.Checker()
+		if err != nil {
+			return err
+		}
+		if succeed {
+			hh.report.Checks = append(hh.report.Checks, api.HealthcheckItem{
+				Name:    li.Name,
+				Status:  api.HealthcheckStatusOK,
+				Message: fmt.Sprintf("%s: OK", li.Name),
+			})
+			continue
+		}
+		// Checker failed, Append the failure to the check report
+		hh.report.Checks = append(hh.report.Checks, api.HealthcheckItem{
+			Name:    li.Name,
+			Status:  api.HealthcheckStatusFailed,
+			Message: fmt.Sprintf("%s: FAILED. Fixing: %t", li.Name, fix),
```

Comment:
We're losing information we output in the current version. This is a bit of a regression. The message is redundant with the Status. Above and here, you output OK/FAILED, and the name of the fix. All that information is already in the struct. The message field is intended for message that explain WHY a healthcheck failed.